org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/corext/template/java/CodeSnippetTemplate.java

```diff
@@ -0,0 +1,92 @@
+	IFNOTNULL(TemplatePreferences.IFNOTNULL_ID, JavaContextType.ID_STATEMENTS, TemplatePreferences.IFNOTNULL_CONTENT, TemplatePreferences.IFNOTNULL_DESCRIPTION);
+	//@formatter:on
+
+	private final String templateId;
+	private final String contextType;
+	private final String defaultContent;
+	private final String description;
+
+	private CodeSnippetTemplate(String templatesId, String contextType, String defaultContent, String description) {
+		this.templateId = templatesId;
+		this.contextType = contextType;
+		this.defaultContent = defaultContent;
+		this.description = description;
+	}
+
+	public Template createTemplate() {
+		return new Template(this.name().toLowerCase(), this.description, this.contextType, this.defaultContent, false);
+	}
+
+	public String getId() {
+		return this.templateId;
+	}
+}
+
+class TemplatePreferences {
+	// IDs
+	public static final String SYSOUT_ID = "org.eclipse.jdt.ls.templates.sysout";
+	public static final String SYSERR_ID = "org.eclipse.jdt.ls.templates.syserr";
+	public static final String SYSTRACE_ID = "org.eclipse.jdt.ls.templates.systrace";
+	public static final String FOREACH_ID = "org.eclipse.jdt.ls.templates.for_array";
+	public static final String FORI_ID = "org.eclipse.jdt.ls.templates.for_iterable";
+	public static final String WHILE_ID = "org.eclipse.jdt.ls.templates.while_condition";
+	public static final String DOWHILE_ID = "org.eclipse.jdt.ls.templates.do";
+	public static final String IF_ID = "org.eclipse.jdt.ls.templates.if";
+	public static final String IFELSE_ID = "org.eclipse.jdt.ls.templates.ifelse";
+	public static final String IFNULL_ID = "org.eclipse.jdt.ls.templates.ifnull";
+	public static final String IFNOTNULL_ID = "org.eclipse.jdt.ls.templates.ifnotnull";
+
+	// DefaultContents
+	public static final String SYSOUT_CONTENT = "System.out.println($${0});";
+	public static final String SYSERR_CONTENT = "System.err.println($${0});";
+	public static final String SYSTRACE_CONTENT = "System.out.println(\"${enclosing_type}.${enclosing_method}()\");";
+	public static final String FOREACH_CONTENT = "for (${iterable_type} ${iterable_element} : ${iterable}) {\n" + "\t$${0}\n" + "}";
+	public static final String FORI_CONTENT = "for (int ${index} = 0; ${index} < ${array}.length; ${index}++) {\n" + "\t$${0}\n" + "}";
+	public static final String WHILE_CONTENT = "while (${condition:var(boolean)}) {\n" + "\t$${0}\n" + "}";
+	public static final String DOWHILE_CONTENT = "do {\n" + "\t$${0}\n" + "} while (${condition:var(boolean)});";
+	public static final String IF_CONTENT = "if (${condition:var(boolean)}) {\n" + "\t$${0}\n" + "}";
+	public static final String IFELSE_CONTENT = "if (${condition:var(boolean)}) {\n" + "\t$${0}\n" + "} else {\n" + "\t\n" + "}";
+	public static final String IFNULL_CONTENT = "if (${name:var} == null) {\n" + "\t$${0}\n" + "}";
+	public static final String IFNOTNULL_CONTENT = "if (${name:var} != null) {\n" + "\t$${0}\n" + "}";
```

Comment:
Looks like we're losing the ability to tab through placeholders, which is pretty inconvenient. Compare with https://github.com/redhat-developer/vscode-java/blob/cab76747ae4b0aeef1abcb42294afd28715e2924/snippets/java.json#L37

Reply:
This works:
```
FOREACH_CONTENT = "for ($${1:${iterable_type}} $${2:${iterable_element}} : $${3:${iterable}}) {\n" + "\t$${0}\n" + "}";
```

Reply:
Good idea, that makes it much better!